### PR TITLE
[Relay][TOPI]Fix meaning of conv2d_transpose output_padding parameter

### PR DESCRIPTION
**Pull request by abergeron**
I've fixed the meaning of output_padding to be more in line with what other machine learning frameworks and libraries intend for the meaning and also to make it actually useful for the gradient.

This changed the definition of conv2d_transpose in TOPI in a minor way but I tried to make the change transparent as much as possible.

@vinx13

### CODE REVIEW

**Comment by vinx13** [comment]
there are some issue with ci, please retriever the ci

**Comment by vinx13** [comment]
We need to update logs in https://github.com/uwsampl/tvm-distro/tree/master/tophub

**Comment by abergeron** [comment]
I don't know how to retrigger CI besides adding a dummy commit.  If there is a better way, please do tell.

**Comment by junrushao1994** [comment]
git commit --allow-empty -m “Trigger CI”

**Comment by vinx13** [comment]
CI is still sad :(
You can also `git commit --amend` to reset the last commit and force push, instead of adding a dummy commit

**Comment by abergeron** [comment]
It looks like I missed the VTA template.  This one doesn't have any tests for the output padding and it makes me slightly uncomfortable, but the only test there is is for a very specific workload so I don't want to mess with it.

**Comment by vinx13** [comment]
> It looks like I missed the VTA template. This one doesn't have any tests for the output padding and it makes me slightly uncomfortable, but the only test there is is for a very specific workload so I don't want to mess with it.

cc @tmoreau89

**Comment by abergeron** [comment]
I would appreciate some help with the VTA failure because it doesn't appear to be related to something i've done (I didn't add any if_then_else nodes), but at the same time it is in something I've modified (the conv2d_transpose test).

So maybe I broke something, but I'm really not sure.

**Comment by tmoreau89** [comment]
@abergeron is the issue right now that VTA test will fail when you set output padding to `(1,1)`? Or does it also fail when it's set to `(0,0)` (current test case)

**Comment by abergeron** [comment]
It fails with (0, 0) in the CI right now.

It might also be a good idea to test it with non-zero values to make sure it works.

**Comment by tmoreau89** [comment]
I see; I'm a little confused because the test case with output padding set to `(0,0)` should be inconsequential; however it seems to break the VTA test. How time critical is this PR? I can try to reproduce the test in your branch over the weekend.

**Comment by abergeron** [comment]
This is not super time critical, so you can take time to make sure it works properly.

But I would like this to be merged within a week or two if possible.

**Comment by tmoreau89** [comment]
@abergeron ack, I'll investigate to see what causes the failure at the moment. Will report back by Tuesday.

**Comment by abergeron** [comment]
@tmoreau89 Any news?

**Comment by tmoreau89** [comment]
@abergeron got sidetracked with other commitments, I'll reproduce the issue tomorrow.

**Comment by tmoreau89** [comment]
FYI, I was able to reproduce the issue in your branch, I'll look more into what's causing it tomorrow...

**Comment by tmoreau89** [comment]
@abergeron: I have good new and bad news. The *good news* is that I was able to find the root cause of the observed bug. By changing the `conv2d_transpose_nchw` function signature by adding the `output_padding` argument, we are essentially changing the TOPHUB schedule lookup convention for that operator. In other words, AutoTVM fails to find the schedule for the conv2dtranspose operator because the argument list of the operator doesn't match. Due to this, the schedule defaults to an illegal one for VTA which triggers the error we're seeing in the CI.

In order to circumvent the error the fix will be straightforward on your end: just change `v0.06` to `v0.07` here: https://github.com/apache/incubator-tvm/blob/master/python/tvm/autotvm/tophub.py#L59

I've updated the schedule for VTA with this commit: https://github.com/uwsampl/tvm-distro/commits?author=tmoreau89

Essentially I changed the DCGAN schedules from 
`["conv2d_transpose_nchw", [1, 64, 4, 4, 1, 16, "int8"], [32, 64, 4, 4, 16, 16, "int8"], [2, 2], [1, 1], "int32"]` 
to 
`["conv2d_transpose_nchw", [1, 64, 4, 4, 1, 16, "int8"], [32, 64, 4, 4, 16, 16, "int8"], [2, 2], [1, 1], "int32", [0, 0]]`

The *bad news* is that this will need to be modified for all hardware targets under tophub. For instance the CUDA schedule will need to change: https://github.com/uwsampl/tvm-distro/blob/master/tophub/cuda_v0.06.log#L690 among other targets. 

In order to do so, I recommend first creating new schedule files under a new "package version" and then switching the package version in your PR in `tophub.py` in order not to break other unit tests. The reason why this is not causing an error is that CPUs and GPUs schedules default to valid, albeit slow ones (VTA is just unstable and not all schedules will lead to correct execution, we need a better defaulting mechanism).

Finally, I also created a branch that will fix some scripts under VTA that will be affected from your change. If you could cherry pick my commits into your branch that would be great: https://github.com/tmoreau89/tvm/tree/4318_fix

**Comment by abergeron** [comment]
I consider this to be 100% good news since I would only have to update the schedules.  I needed to figure out how to use autotvm for other purposes soon so I'll get on that.

**Comment by tmoreau89** [comment]
Great then! Happy to provide you with guidance as you go through the changes in tophub.

**Comment by abergeron** [comment]
For now, I've included your fixes and rebase on the current master.  I'll do the tuning for non-VTA targets later.

**Comment by tmoreau89** [comment]
Cool, let's see if the CIs pass; in the meantime, I don't think it's necessary to tune the targets, it's just a matter of changing the entries to include the new operator argument

**Comment by tmoreau89** [comment]
In the case of VTA, I had to add the new field in the argument field as to not break the schedule lookup mechanism:

from
["conv2d_transpose_nchw", [1, 64, 4, 4, 1, 16, "int8"], [32, 64, 4, 4, 16, 16, "int8"], [2, 2], [1, 1], "int32"]
to
["conv2d_transpose_nchw", [1, 64, 4, 4, 1, 16, "int8"], [32, 64, 4, 4, 16, 16, "int8"], [2, 2], [1, 1], "int32", **[0, 0]**]

**Comment by abergeron** [comment]
I discovered that backends other than VTA don't have convenient scripts to do the tuning.  I will probably write some so that we have a script in the repo that can reproduce tuning files in tophub.

**Comment by tmoreau89** [comment]
Hmmm, is the reason you want to re-tune schedules to cover cases where output pad = `(1,1)`?

**Comment by abergeron** [comment]
I forgot the tophub.py file update because I thought it was in your branch.

Also, If I don't really need to retune anything, then I would rather save the time.  But there might have been some output padding in the model that uses conv2d_transpose on tophub.  I'm not sure exactly what model is that however.

In any case I think it would be very good to have a reference script that produces the tuning files in tophub, so it would be easy to recreate the files when something changes.

**Comment by abergeron** [comment]
I've tried all the models in the benchmark files (in apps/benchmarks) and none of them use any convolution.  So I have no idea where the conv2d_transpose in the tuning files comes from.

If someone knows where those come from, it would help a lot.  Also I think this kinds of reinforces the idea that a publicly available script to make those files would help a lot, unless it is already present somewhere I missed.

**Comment by vinx13** [comment]
@abergeron The original logs for conv2d_transpose in top hub assumes zero output_padding. So you can just update the logs to add output_padding in the workload so that the query after this pr can be matched

**Comment by abergeron** [comment]
I've made https://github.com/uwsampl/tvm-distro/pull/13 which *should* be ok

**Comment by tmoreau89** [comment]
Thanks @abergeron; I've merged the changes into tvm-distro. Did you spot-check that the conv2d_transpose operator schedule lookup work correctly in topi? If so, you can update the version number in tophub and this PR should be good to go.

**Comment by abergeron** [comment]
Strangely it seems that topi never loads anything for conv2d_transpose.  So I'm not sure what to spot check.

**Comment by vinx13** [comment]
@abergeron You can test compile a `conv2d_transpose` in Relay and check if the log is loaded

**Comment by abergeron** [comment]
I'll do that.

**Comment by abergeron** [comment]
I don't see anything loading for conv2d_transpose in my test code, but it also never seems to load anything at all.  Did I miss a codepath that does the loading?

**Comment by vinx13** [comment]
@abergeron I tested locally and it looks good to me. If your code is not called, this is likely generic func has been registered for other targets. For example, in llvm target, `x86/conv2d_transpose` will be called instead of `nn/conv2d_transpose`. Could you rebase so that we can get this in?

**Comment by abergeron** [comment]
I did a rebase, but in the progress I discovered that someone added conv1d with the same problem as conv2d for output_padding.  I'll fix that too.

File: python/tvm/autotvm/tophub.py [hunk 1]
```diff
@@ -55,7 +55,7 @@
     'mali':             "v0.05",
     'intel_graphics':   "v0.01",
 
-    'vta':              "v0.06",
+    'vta':              "v0.07",
```
Comment (vinx13) [inline python/tvm/autotvm/tophub.py]: cuda and arm cpu version also need to be updated

Comment (abergeron) [reply python/tvm/autotvm/tophub.py]: done

**Comment by vinx13** [comment]
@abergeron Thanks for the fix

**Comment by icemelon9** [comment]
@abergeron @vinx13 @tmoreau89 
I found out two problems in this PR. 
1. In this [line](https://github.com/apache/incubator-tvm/pull/4318/files#diff-8be7003a84f663f3f6b0dbb3bf1f5ba6R105), `h+dh` can be potentially out of boundary. `max(h) = out_h - 1 = in_h - filter_h + output_padding[0]` and `max(dh) = filter_h - 1`, therefore, `max(h+dh) = in_h + output_padding[0] - 1`. When `output_padding[0] >= 1`, `max(h+dh) >= in_h`, which is out of `data_pad` height boundary. Similar to `w+dw`.
2. The x86 conv2d_transpose implementation is different from generic conv2d_tranpose. In x86, after you call `conv2d_transpose_nchw_preprocess`, you directly call the normal conv2d without using `output_padding`.

I'll revert this PR for now. Could you fix these two bugs and double check whether cuda and arm_cpu are correct? Further, could you investigate why CI doesn't catch these errors?
